src/shims/unix/linux/epoll.rs

```diff
@@ -411,12 +411,13 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         let mut num_of_events: i32 = 0;
         let mut array_iter = this.project_array_fields(&event)?;
 
-        while let Some((epoll_key, epoll_return)) = ready_list.pop_first() {
-            // If the file description is fully close, the entry for corresponding FdID in the
-            // global epoll event interest table would be empty.
-            if this.machine.epoll_interests.get_epoll_interest(epoll_key.0).is_some() {
-                // Return notification to the caller if the file description is not fully closed.
-                if let Some(des) = array_iter.next(this)? {
+        while let Some(des) = array_iter.next(this)? {
+            // Fetch an event from the ready list.
+            if let Some((epoll_key, epoll_return)) = ready_list.pop_first() {
```

Comment:
This will cause empty entries whenever a FD was already closed. You'll need to use a for loop here and exit it only if you actually write an entry.

This will cause us to finish iterating over the array_iter and never doing anything once the ready list is empty. That's fine, but we could also ensure we `continue` to the outer loop (by giving it a label) and unconditionally breaking after the ready list loop

Reply:
(fwiw: this is why we abandoned this approach before, but forgot that in abandoning it we'd not fix this issue)

Reply:
Indeed, this needs to be a *loop* getting the next entry from the ready list that refers to an FD that is still open.

(I assume there's a test covering this case? That test would generate an interesting event on an FD, and then close the FD, and only then call epoll_wait and expect to *not* see that event.)

Reply:
Right, the test that I added initially failed because there is empty entry where it shouldn't. I added a condition to check if the entry is actually written, also terminate earlier if the ready list is empty (although this won't affect correctness).